### PR TITLE
[Channel] Brand new channel context!

### PR DESCRIPTION
**Pull request by pamil**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Fixed tickets | #3929
| License       | MIT
| Doc PR        | -

Reworking the channel context, introducing request based channel context that uses request resolvers.

### CODE REVIEW

File: src/Sylius/Component/Channel/Context/RequestBased/ChannelContext.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        return $channel;
+    }
+
+    /**
+     * @return Request
+     */
+    private function getMasterRequest()
+    {
+        $masterRequest = $this->requestStack->getMasterRequest();
+        if (null === $masterRequest) {
```
Comment (tristanbes) [inline src/Sylius/Component/Channel/Context/RequestBased/ChannelContext.php]: you should break a new line before the if statement for better redability

File: src/Sylius/Component/Channel/Context/RequestBased/CompositeRequestResolver.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class CompositeRequestResolver implements RequestResolverInterface
+{
+    /**
+     * @var PriorityQueue|RequestResolverInterface[]
+     */
+    private $priorityQueue;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Channel/Context/RequestBased/CompositeRequestResolver.php]: `$priorityQueue` -> `$requestResolvers`?

File: src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php [hunk 3]
```diff
@@ -24,28 +22,8 @@ class ChannelRepository extends EntityRepository implements ChannelRepositoryInt
...
-            ->setParameter('hostname', '%'.$hostname.'%')
-        ;
-
-        return $queryBuilder->getQuery()->getOneOrNullResult();
-    }
-
-    /**
-     * {@inheritdoc}
-     */
-    public function findDefault()
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php]: It would be good to support single channel without need to configure hostname or anything else, maybe we should have a default channel? We could have the default channel resolver that will be the last resolver and will return the channel that is marked as default.

Comment (aramalipoor) [inline src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php]: :+1:

Comment (pjedrzejewski) [inline src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php]: Let's discuss this today, sounds reasonable, but can lead to some unexpected behavior, especially when poorly documented.

Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php]: :+1:, but it should be optional and made at `ChannelContext` level as it does not involve any request-related logic. Anyway, I think we can agree that finding default channel is not the responsibility of channel repository.

File: src/Sylius/Component/Channel/Context/RequestBased/CompositeRequestResolver.php (hunk 2 above)
Comment (pamil) [inline src/Sylius/Component/Channel/Context/RequestBased/CompositeRequestResolver.php]: :+1:

File: src/Sylius/Component/Channel/Context/RequestBased/ChannelContext.php (hunk 1 above)
Comment (pamil) [inline src/Sylius/Component/Channel/Context/RequestBased/ChannelContext.php]: It's a common practice not to put a blank line between those if one line is assigning the value to A and the second one is `if` doing some comparison using A.

File: src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php (hunk 3 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/Doctrine/ORM/ChannelRepository.php]: Okay, there it is - `Sylius\Component\Channel\Context\SingleChannel\ChannelContext` :)

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        if (!$container->hasDefinition('sylius.context.channel.request_based.resolver')) {
+            return;
+        }
+
+        $requestResolverDefinition = $container->findDefinition('sylius.context.channel.request_based.resolver');
+
+        $taggedServices = $container->findTaggedServiceIds('sylius.context.channel.request_based.resolver');
+        foreach ($taggedServices as $id => $tags) {
+            foreach ($tags as $attributes) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: Won't it be more readable if you extract it to private method?

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+namespace Sylius\Bundle\ChannelBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class CompositeRequestResolverPass implements CompilerPassInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: Why do you used final here?

File: src/Sylius/Component/Channel/composer.json [hunk 6]
```diff
@@ -23,7 +23,8 @@
         "php": "^5.5.9|^7.0",
 
         "doctrine/orm": "^2.4.8,<2.5",
-        "sylius/resource": "0.17.*"
+        "sylius/resource": "0.17.*@dev",
```
Comment (pamil) [inline src/Sylius/Component/Channel/composer.json]: To be reverted

File: src/Sylius/Component/Channel/composer.json [hunk 7]
```diff
@@ -37,7 +38,6 @@
     "autoload-dev": {
         "psr-4": { "Sylius\\Component\\Channel\\spec\\": "spec/" }
     },
-    "minimum-stability": "dev",
```
Comment (pamil) [inline src/Sylius/Component/Channel/composer.json]: Minimum stability + prefer stable

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php (hunk 5 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: There's no need to extend this service, I haven't ever seen a use-case where extending compiler pass would be a solution.

File: src/Sylius/Component/Channel/composer.json (hunk 7 above)
Comment (lchrusciel) [inline src/Sylius/Component/Channel/composer.json]: Why did you remove this? Did we change convention?

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeChannelContextPass.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class CompositeChannelContextPass implements CompilerPassInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition('sylius.context.channel')) {
```
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeChannelContextPass.php]: TODO: Define that service at `sylius.context.channel.composite` instead and proceed only if `sylius.context.channel` does not exist. If someone would like to provide different implementation, we don't want to add method calls on another service. At the end of the function the alias `sylius.context.channel` -> `sylius.context.channel.composite` should be set

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php (hunk 5 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: :+1:

File: src/Sylius/Component/Channel/Context/SingleChannel/ChannelContext.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+namespace Sylius\Component\Channel\Context\SingleChannel;
+
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Channel\Repository\ChannelRepositoryInterface;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class ChannelContext implements ChannelContextInterface
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Channel/Context/SingleChannel/ChannelContext.php]: Not not renaming to `SingleChannelContext` and putting it in `Sylius\Component\Channel\Context` namespace? IMO it's overhead to create a separate namespace for that :)

File: src/Sylius/Component/Channel/spec/Context/RequestBased/CompositeRequestResolverSpec.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        $this->shouldHaveType('Sylius\Component\Channel\Context\RequestBased\CompositeRequestResolver');
+    }
+
+    function it_implements_request_resolver_interface()
+    {
+        $this->shouldImplement(RequestResolverInterface::class);
+    }
+
+    function it_returns_null_if_there_are_no_nested_request_resolvers_added(
+        Request $request
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Channel/spec/Context/RequestBased/CompositeRequestResolverSpec.php]: No need to put it in a separate line :)

**Comment by michalmarcinkowski** [comment]
Well done Kamil! :+1:

File: src/Sylius/Component/Channel/Context/SingleChannel/ChannelContext.php (hunk 9 above)
Comment (pamil) [inline src/Sylius/Component/Channel/Context/SingleChannel/ChannelContext.php]: :+1:

File: src/Sylius/Component/Channel/Context/RequestBased/CompositeRequestResolver.php [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findChannel(Request $request)
+    {
+        foreach ($this->requestResolvers as $requestResolver) {
+            $channel = $requestResolver->findChannel($request);
+
+            if (null === $channel) {
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Channel/Context/RequestBased/CompositeRequestResolver.php]:
```
if (null !== $channel) {
    return $channel;
}
```

File: src/Sylius/Component/Channel/spec/Context/SingleChannelContextSpec.php [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    function it_returns_a_channel_if_it_is_the_only_one_defined(
+        ChannelRepositoryInterface $channelRepository,
+        ChannelInterface $channel
+    ) {
+        $channelRepository->findAll()->willReturn([$channel]);
+
+        $this->getChannel()->shouldReturn($channel);
+    }
+
+    function it_throws_an_channel_not_found_exception_if_there_are_no_channels_defined(
```
Comment (lchrusciel) [inline src/Sylius/Component/Channel/spec/Context/SingleChannelContextSpec.php]: Won't it be better to test all exception throws in a one function?

Comment (pamil) [inline src/Sylius/Component/Channel/spec/Context/SingleChannelContextSpec.php]: There are the same behaviours expected from different backgrounds, specifications should say explicitly when something happens, not only that "it_throws_an_exception".

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+class CompositeRequestResolverPassTest extends AbstractCompilerPassTestCase
+{
+    /**
+     * @test
+     */
+    public function it_collects_tagged_request_based_channel_contexts()
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: Do we have any convention defined for phpUnit tests? IIRC we should use camel case instead of snake case or am I missed something?

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/PrioritizedCompositeServicePass.php [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        if (!$container->hasDefinition($this->compositeId)) {
+            return;
+        }
+
+        $channelContextDefinition = $container->findDefinition($this->compositeId);
+
+        $taggedServices = $container->findTaggedServiceIds($this->tagName);
+        foreach ($taggedServices as $id => $tags) {
+            foreach ($tags as $attributes) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/PrioritizedCompositeServicePass.php]: Won't it be cleaner to extract this for each to separate method?

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Sylius\Bundle\ChannelBundle\DependencyInjection\Compiler;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class CompositeRequestResolverPass extends PrioritizedCompositeServicePass
+{
+    /**
+     * {@inheritdoc}
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: Do we need this inheritdoc if constructor doesn't need any parameters?

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php (hunk 13 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: I don't know, I used the same convention as in PhpSpec because snake_case makes reading long sentences easier that camelCase.

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/PrioritizedCompositeServicePass.php (hunk 14 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/PrioritizedCompositeServicePass.php]: :+1:

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php (hunk 15 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: In fact I'm not sure here - on the one hand it does not need these arguments, on the other - I consider `{@inheritdoc}` like Java's `@Override`, just as a mark that the method overrides parent method.

File: src/Sylius/Component/Channel/spec/Context/SingleChannelContextSpec.php (hunk 12 above)
Comment (lchrusciel) [inline src/Sylius/Component/Channel/spec/Context/SingleChannelContextSpec.php]: I'm totally agree for above case in [CompositeChannelContextSpec](https://github.com/Sylius/Sylius/pull/4056/files#diff-cd860143795fd6ba7770e844de50c2ceR37). But here we have just different values returned by repository function. Won't it be better?

```php
    function it_throws_an_channel_not_found_exception_if_there_isnt_one_channel_defined(
        ChannelRepositoryInterface $channelRepository
    ) {
        $channelRepository->findAll()->willReturn([], [$firstChannel, $secondChannel]);

        $this->shouldThrow(ChannelNotFoundException::class)->during('getChannel');
        $this->shouldThrow(ChannelNotFoundException::class)->during('getChannel');
    }
```

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php (hunk 13 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: You are right, but on the other hand shouldn't we keep consistency? At the end tests in theme bundle or api tests are written with camel case. Should we force some convention?

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php (hunk 15 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: I'm not convinced about that. For example in Product class in core bundle we use parent constructor, but don't use `{@inheritdoc}` tag.

File: src/Sylius/Component/Channel/spec/Context/SingleChannelContextSpec.php (hunk 12 above)
Comment (pamil) [inline src/Sylius/Component/Channel/spec/Context/SingleChannelContextSpec.php]: IMHO it makes it much more difficult to read, if I can make two simple specifications instead of one complex, I'll go for the first option.

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php (hunk 13 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: Right now we does not regulated any convention, ThemeBundle has already two of them inside (camelCase in functional, snake_case in configuration). If there will be any, I'll follow it, but for now I'll let this as is and vote for snake_case.

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php (hunk 15 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: All right, it's not a big deal.

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ChannelBundle\DependencyInjection\Compiler;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class CompositeRequestResolverPass extends PrioritizedCompositeServicePass
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: This based class is pretty generic, later we should think about appropriate place for that. We use so much stuff from Symfony DIC that we could have a standalone Sylius DependencyInjection component with all this crazy stuff haha. :)

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php (hunk 13 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: I did snake_case in #3936 and I think we should use it because it is more readable. That being said, I am not sure how that will work out for API tests in PHPUnit:

```
function it_allows_creating_product()
function it_denies_access_to_product_creation_for_not_authenticated_users()
function it_returns_not_found_response_for_non_existing_product()
function it_creates_customer_without_user_account()
function it_allows_to_delete_users()
function it_allows_to_update_users()
```

Thoughts?

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+            'sylius.context.channel.request_based.resolver',
+            'addResolver',
+            [new Reference('sylius.context.channel.request_based.resolver.tagged_one'), 42]
+        );
+    }
+
+    /**
+     * @test
+     */
+    public function it_does_not_add_method_calls_to_overwriting_service_if_the_composite_service_is_overwrited()
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: overwritten or overridden?

File: src/Sylius/Bundle/ChannelBundle/composer.json [hunk 18]
```diff
@@ -27,7 +27,9 @@
         "symfony/framework-bundle": "^2.7.7"
     },
     "require-dev": {
+        "matthiasnoback/symfony-dependency-injection-test": "^0.7.6",
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ChannelBundle/composer.json]: Wouldn't be more appropriate to have this next to PHPUnit? We do not sort deps alphabetically.

File: src/Sylius/Bundle/ChannelBundle/composer.json [hunk 19]
```diff
@@ -41,6 +43,7 @@
         "psr-4": { "Sylius\\Bundle\\ChannelBundle\\spec\\": "spec/" }
     },
     "minimum-stability": "dev",
+    "prefer-stable": true,
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ChannelBundle/composer.json]: Why? Symfony components/bundles do not use that flag.

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php (hunk 13 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: It can be even better as test names for API can be pretty long. :+1: I will adapt #4037 to snake case.

File: src/Sylius/Bundle/ChannelBundle/composer.json (hunk 19 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/composer.json]: Installing that bundle standalone without prefer-stable downloads a lot of non-stable dependencies. If we want to test bundles standalone, I believe that it is the way to go. 

It does not affect any packages using it as these configuration options - "minimum-stability" and "prefer-stable" - are considered only for root packages. Sylius dependencies requiring `0.17.*` still resolves to the latest, development version.

File: src/Sylius/Bundle/ChannelBundle/composer.json (hunk 18 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/composer.json]: IIRC Sylius' main `composer.json` is sorted in that way, so I decided to stick to that.

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php (hunk 17 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: Ooops, something has gone wrong! :)

File: src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php (hunk 16 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/DependencyInjection/Compiler/CompositeRequestResolverPass.php]: Yep, the same with dynamic forms in metadata and so on...

File: src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php (hunk 13 above)
Comment (pamil) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: Awesome :+1:

Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ChannelBundle/Tests/DependencyInjection/Compiler/CompositeRequestResolverPassTest.php]: :+1: 

Just to clarify, in order to stick to the convention with 'allow' we should rename:
`function it_creates_customer_without_user_account()`
to:
`function it_allows_to_create_customer_without_user_account()`

File: src/Sylius/Bundle/ChannelBundle/composer.json (hunk 19 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ChannelBundle/composer.json]: Ok, I opened RFC to add this to all bundles/components then - https://github.com/Sylius/Sylius/issues/4088.

**Comment by pjedrzejewski** [comment]
Awesome work Kamil! :+1:
